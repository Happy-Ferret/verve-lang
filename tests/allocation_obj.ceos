type BigObject {
  BigObject(String, String, String, String, String, String, String, String) // 8 Strings
}

type Pair {
  Pair(Int, Int)
}

type Wrapper {
  Wrapper(Pair)
}

allocate(): BigObject {
  BigObject(
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo"
  )
}

a(): BigObject { allocate() allocate() }
b(): BigObject { a() a() }
c(): BigObject { b() b() }
d(): BigObject { c() c() }
e(): BigObject { d() d() }
f(): BigObject { e() e() }
g(): BigObject { f() f() }
h(): BigObject { g() g() }
i(): BigObject { h() h() }
j(): BigObject { i() i() }
k(): BigObject { j() j() }
l(): BigObject { k() k() }
m(): BigObject { l() l() }
n(): BigObject { m() m() }

test(wrapped: Wrapper): Void {
  n() // allocate a lot and trigger garbage collection
  let Wrapper(pair) = wrapped {
    let Pair(a, b) = pair { 
      print([a, b])
    }
  }
}

test(Wrapper(Pair(42, 13)))

// this will allocate 4243744b (~4mb)
// the GC should get it down to ~20k
// here we just test it's less than ~30k
print(__heap-size__() < 30000)
