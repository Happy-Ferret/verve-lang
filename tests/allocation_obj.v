type big_object {
  BigObject(string, string, string, string, string, string, string, string) // 8 strings
}

type pair {
  Pair(int, int)
}

type wrapper {
  Wrapper(pair)
}

fn allocate() -> big_object {
  BigObject(
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo",
    "foooooooooooooooooooooooooooooooooooooooooooo"
  )
}

fn a() -> big_object { allocate() allocate() }
fn b() -> big_object { a() a() }
fn c() -> big_object { b() b() }
fn d() -> big_object { c() c() }
fn e() -> big_object { d() d() }
fn f() -> big_object { e() e() }
fn g() -> big_object { f() f() }
fn h() -> big_object { g() g() }
fn i() -> big_object { h() h() }
fn j() -> big_object { i() i() }
fn k() -> big_object { j() j() }
fn l() -> big_object { k() k() }
fn m() -> big_object { l() l() }
fn n() -> big_object { m() m() }

fn test(wrapped: wrapper) -> void {
  n() // allocate a lot and trigger garbage collection
  let Wrapper(pair) = wrapped {
    let Pair(a, b) = pair { 
      print([a, b])
    }
  }
}

test(Wrapper(Pair(42, 13)))

// this will allocate 4243744b (~4mb)
// the GC should get it down to ~20k
// here we just test it's less than ~30k
print(`__heap-size__`() < 30000)
