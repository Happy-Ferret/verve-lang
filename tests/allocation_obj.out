42 13
1
